content/blog/teach-tidyverse-2021/index.Rmd

```diff
@@ -0,0 +1,437 @@
+
+1.  Similar interfaces to different models.
+2.  Model outputs as tibbles, which are straightforward to interact with for learners who already know how to wrangle and visualize data stored in this format.
+3.  Features that help users avoid common machine learning pitfalls.
+
+Let's start with the first one --- providing similar interfaces to models.
+Consider the question "*How do you define the the number of trees when fitting a random forest model?"* The answer is generally *"depends on the package: `randomForest::randomForest()` uses `ntree`, `ranger::ranger()` uses `num.trees`, Spark's `sparklyr::ml_random_forest()` uses `num_trees`"*.
+The answer with tidymodels is a bit simpler though: *"using the `trees` argument in the `rand_forest()` package, regardless of the engine being used to fit the model"*.
+This can allow new learners to focus on what "trees" mean and how one decides how many to use, instead of the precise syntax needed by the various packages that can fit random forest models.
+
+The pedagogical advantages of teaching modeling with the full tidymodels framework may not be clear for fitting simple models with `lm()`.
+For example, below we fit a simple linear regression model with a single predictor, using base R first and then using tidymodels.
+
+```{r}
+# base R
+lm(hwy ~ cty, data = mpg) %>%
+  summary()
+
+# tidymodels
+linear_reg() %>%
+  set_engine("lm") %>%
+  fit(hwy ~ cty, data = mpg) %>%
+  tidy()
+```
+
+The tidymodels approach takes a few more steps, and for a simple model like this, the only advantage is likely in the summarisation step.
+With `tidy()`, we get the model output as a tibble, which is more straightforward to interact with programmatically and which, by default, omits the significant stars.
+
+```{r}
+lm(hwy ~ cty, data = mpg) %>%
+  tidy()
+```
+
+The pedagogical advantages for the consistent API of the framework become more clear when we move on to fitting different models.
+Below you can see examples of how we can fit models using various engines or using the same engine, but different modes.
+
+```{r eval = FALSE}
+# different engines
+linear_reg() %>%
+  set_engine("lm") %>%
+  set_mode("regression")
+
+logistic_reg() %>% 
+  set_engine("glm") %>%
+  set_mode("classification")
+
+rand_forest() %>% 
+  set_engine("ranger") %>% 
+  set_mode("regression")
+
+decision_tree() %>% 
+  set_engine("rpart") %>% 
+  set_mode("regression")
+
+# same engine, different modes
+svm_linear() %>% 
+  set_engine("LiblineaR") %>% 
+  set_mode("regression")
+
+svm_linear() %>% 
+  set_engine("LiblineaR") %>% 
+  set_mode("classification")
+```
+
+I should note that fitting a bunch of models to the same data and picking the one you like the results of the best is not a good approach.
+So one would rarely see code as it appears in the chunk above in a single R script.
+Students learning about these models would likely encounter these pipelines over the course of a semester, each in a slightly different data context.
```

Comment:
```suggestion
Fitting a bunch of models to the same data and picking the one you like the results of the best is not a good approach, so one would rarely see code as it appears in the chunk above in a single R script.
Students will encounter these pipelines over the course of a semester, each in a slightly different data context.
Because the syntax is uniform, it's easier to focus on the details of the model, not how to fit the darn thing in R.
```